### ISSUE TITLE
ZFS support

### ISSUE THREAD

**Issue by ghost** [calamares/calamares#533]
Jira issue originally created by user juniocalu:

Hi,
There Any predictions for the introduction of support ZFS?

**Comment by wxl** [comment]
With Ubuntu adding experimental ZFS support in 19.10 including [installer support for creating a root filesystem with it](https://wiki.ubuntu.com/EoanErmine/ReleaseNotes#ZFS_on_root), I'm inclined to check in again regarding the status of this. I don't see any upstream issues which makes me sort of presume (most likely incorrectly) that KPMCore considers their work complete, or at least good enough.

**Comment by monetaryabyss** [comment]
We have recently become aware of some [additions to KPMCore](https://github.com/KDE/kpmcore/search?q=zfs&type=Code) so we wanted to check in on this issue again. Our hope for Lubuntu is to be able to include this for our 20.04 (LTS) release if possible.

**Comment by arcolinuxz** [comment]
Had a user requiring about this feature...
Seems kpmcore made changes.

**Comment by SeveredinTwain** [comment]
Following as well.  I would love to be able to use Calamares to build an ArchRootonZFS custom iso.

**Comment by abucodonosor** [comment]
I don't think that will happen any time soon, considering the licensing issues in Linux.

Also before we support out-of-tree modules with License issues ( and note ZFS support need an explicitly partitioning UI modules), I think we should add the in-kernel BTRFS support, which needs the same, and own UI partitioning modules.

**Comment by adriaandegroot** [comment]
I have a lot more interest in, and sympathy for, ZFS than btrfs. But that's all moot as long as there are more issues coming in than PR's.

**Comment by abucodonosor** [comment]
@adriaandegroot sure, but there is no Licensing problem in BSD.

**Comment by Advantaged** [comment]
Hi there, I also very interested on ZFS.
In attachment you will find (I think a python.script from Ubuntu-installer) `/usr/share/ubiquity/zsys-setup`, the interesting part start at line 316 (if of interest). And here the modified `zpool create`:

> ### Step 2. [Modify configuration file](https://pov.es/linux/ubuntu/ubuntu-20-04-install-ubuntu-with-zfs-and-encryption/)
> 1. Modify `zsys-setup` configuration-file.
> * This here is around line 316
> * Replace THIS-IS-MY-PASSWORD with the encryption password you want to use. You will be prompted to type this at boot. Use here preferably you Login password.
```
# Pools
        # rpool
        echo this-is-my-password | zpool create -f \
                -o ashift=12 \
                -O compression=lz4 \
                -O acltype=posixacl \
                -O xattr=sa \
                -O relatime=on \
                -O normalization=formD \
                -O mountpoint=/ \
                -O canmount=off \
                -O dnodesize=auto \
                -O sync=disabled \
                -O recordsize=1M \
                -O encryption=aes-256-gcm \
                -O keylocation=prompt \
                -O keyformat=passphrase \
                -O mountpoint=/ -R "${target}" rpool "${partrpool}"
```

[zsys-setup.txt](https://github.com/calamares/calamares/files/5758722/zsys-setup.txt)
I backup the whole "Ubiquity"-folder if of interest.

Well, I used Kubuntu for a while but I don't like Ubuntu at all, installing KDE package on top was not my favorite solution but Manjaro-KDE is great.

I'm not an expert but i would help how and where i can, just let me know

**Comment by Advantaged** [comment]
The Arch-Wiki don't helped me so much saying "ZFS is strait forward in Arch", so I try to understand at first the layer of ZFS. From newbie explanation (of FreeBSD) appear:
1. GPT table `gpart create -s gpt da0`
2. Partition `gpart add -t freebsd-zfs -a 4K -l (serial-number-here) da0`
But this is for data-devices and will for sure not works on Manjaro or Arch.

My "problems":
1. I would understand the data-carrier-layers e.g.:
   1.1. Drive-controller and firmware
   1.2. GPT partition table
   1.3. zpools for EFI, Swap and Root?
My question here: Are the "partitions" EFI, Swap and Root "data-sets" formatted respectively to FAT32, linux-swap and ext4?
2. ZFS module installation didn't works on newest kernels even from AUR or github didn't works, with 5.10 not at all, with 5.9 partially (without DKMS, here is a patch that I can't apply), probably 5.8 or 5.7.
Here is the question: with which kernel are the ZFS-module installable and working well?
3. For Grub are many modification required, can we use 'rEFInd'? Are there specific modification to start encrypted ZFS?
4. I don't know how Ubuntu it make to works with `/swapfile' but we should consider a `linux-swap` partition.

Can someone help me for creating at least a good guide for ZFS until the coders make all these feature automated?

**Comment by abucodonosor** [comment]
@Advantaged 

this is the wrong place to ask for ZFS HOWTO's.

**Comment by Advantaged** [comment]
[@ abucodonosor](https://github.com/abucodonosor)

what can I do to help we all?

**Comment by Advantaged** [comment]
The `zsys-setup.sh` (above) is already good, naturally must be modified for Manjaro for creating and use `linux-swap` according to user-choice (normal-RAM, hibernate-RAM).
I would prefer the `alez`-method putting the Swap at end of the disk.

So, I'm writing (first for myself) instructions `Arch-ZFS_prepare-ISO.md` I will add if requested/useful later on when finish, but I have also to point some topics out.
1. The latest Kernel don't works See [archzfs-repo](https://archzfs.com/archzfs/x86_64/) and this should be downgraded mostly, I used 5.10.9. The LTS Kernel is not a solution. Only after downgrading the Kernel as per repo and after add a.m. repo is possible to install `zfs-linux` or `zfs-dkms` from the same. Additionally must be build a control mechanism that upgrade `linux linux-headers zfs-linux` according to version/date of "archzfs-repo"
2.  The `archiso` and `mkinitcpio-archiso` is another delicate topic.
First, (apparently) not all dependencies are installed by default, See [archiso-wiki](https://gitlab.archlinux.org/archlinux/archiso/-/blob/master/docs/README.build). Maybe  is my fold but without Code:
```
pacman -S git make arch-install-scripts edk2-shell squashfs-tools libisoburn disomaster dosfstools lynx mkinitcpio-nfs-utils nbd curl --needed
```
is not possible to install a working `archiso` even if `base-devel` is already present because `/usr/share/archiso/configs/releng/build.sh` is missing.
Second, the actual offered version '50' should be installed first and than downgraded to version '47'. This's also the version of Gitlab which one is linked by the [archlinux_archiso-wiki](https://wiki.archlinux.org/index.php/archiso#Installing_packages).
I can't determine if also for this app is necessary to build a control mechanism for the update.

Finally, we have to solve these (for the moment) two little problems before we start thinking to convert a `zsys-setup.sh` for manjaro.

**Comment by Advantaged** [comment]
> I don't think that will happen any time soon, considering the licensing issues in Linux.
> 
> Also before we support out-of-tree modules with License issues ( and note ZFS support need an explicitly partitioning UI modules), I think we should add the in-kernel BTRFS support, which needs the same, and own UI partitioning modules.
This is not necessary, linux support `ls -la /dev/disk/by-id/` as well the partitioning tolls supports `$DISK`-variable should be defined before e.g. NVMe `DISK=/dev/disk/by-id/nvme-eui.0123456789abcdef`, e.g. SATA-SSD `DISK=/dev/disk/by-id/wwn-0x0123456789abcdef`, e.g. SATA-HDD `DISK=/dev/disk/by-id/wwn-0x0123456789abcdef`. "wwn" is the world-wide-number as well "nvme.eui"
Other variables are e.g.:
```
DISK=/dev/disk/by-id/nvme-eui.0123456789abcdef
POOL=zarchpool-0001
HOST=zarch-0001
USER=terence64w
```

Using examples:
```
blkdiscard $DISK
sgdisk --zap-all $DISK
sgdisk $DISK
blkdiscard $DISK
sgdisk -n 1:1M:+512M -t 1:ef00 -c 1:EFI $DISK
sgdisk -n 2:0:+2G -t 2:8300 -c 2:Boot $DISK
sgdisk -n 3:0:0 -t 3:8309 -c 3:Zub-20-04 $DISK
sgdisk --print $DISK
mkfs.ext4 $DISK-part3
```
Even `/etc/fstab` support `/dev/disk/by-id/wwn-0x0123456789abcdef` as well "rEFInd" as well `mount`, etc.. The only way to find and address  one single disk on large amount of disks in Linux as well in BSD is to use `/dev/disk/by-id/` and/or variable `$DISK`.

Naturally we can setup ZFS using `/dev/sda` etc. but before using it we need to export `zfs zpool-export` (unmount) and reimport `zfs zpool-import` (remount ZFS, about the exact command I'm not sure) it because by import/reimport than ZFS use the ID.

By the way, `zsys-setup.sh` use it or disk-variables as well other variables, these are implemented in scripts used by 'ubiquity', 'calamares', 'alez` and all others OS-install-apps.

We can do it, help me for helping you. I'm not a coder but use now since more than ten years linux.

Please organize a 'Telegram' or 'Signal' group, a 'Forum' or anything else for discuss, organize, schedule next steps.

**Comment by dalto8** [comment]
> I have a lot more interest in, and sympathy for, ZFS than btrfs. But that's all moot as long as there are more issues coming in than PR's.

Since zfs support seems to have been added to kpmcore, would there be support for zfs in Calamares if someone were to do the work and submit a PR?

**Comment by Advantaged** [comment]
Well, apparently GRUB have also problems with ZFS and must go.
A distro-dedicated [rEFInd](http://omniverse.artixlinux.org/x86_64/) in case wished and the `Zfsbootmenu` [here](https://awesomeopensource.com/projects/bootloader) and [here](https://awesomeopensource.com/project/zbm-dev/zfsbootmenu)  should give everyone the possibility to boot a still working kernel and/or snapshot without to know complicated terminal commands.

Besides, some NVMe-manufacturers (Corsair serie [MP510, MP600]) offer the possibility to switch their device to real 4k (4096 bit) PBS/LBS (physical/logical block size) by just one command that could also be implemented in "calamres".

At end, even the non expert user should be able to install and use ZFS without putting even one line of code in the terminal.

Snapshots should be created by each update and/or installation of software and give everyone the possibility to rollback without letting any trace on data-carrier.

**Comment by dalto8** [comment]
> Well, apparently GRUB have also problems with ZFS and must go.

grub can work with zfs with the proper configuration but this doesn't seem like the right place to discuss that.

I started looking at what it would take to add support for zfs in calamares.  There are a lot of things that probably need to be modified.  In addition to the obvious things like mount options and filesystem creation it requires bootloader changes, mkinitcpio, dracut, and encryption changes and probably more when you really get into it.

**Comment by kkofler** [comment]
> Besides, some NVMe-manufacturers (Corsair serie [MP510, MP600]) offer the possibility to switch their device to real 4k (4096 bit) PBS/LBS (physical/logical block size) by just one command that could also be implemented in "calamres".

Not only does this part have essentially nothing to do with ZFS, but this is also not something I would expect an installer to do on a running system. IMHO, it belongs into a kernel CLI option.

**Comment by Advantaged** [comment]
> > Well, apparently GRUB have also problems with ZFS and must go.
> 
> grub can work with zfs with the proper configuration but this doesn't seem like the right place to discuss that.
> 
> I started looking at what it would take to add support for zfs in calamares. There are a lot of things that probably need to be modified. In addition to the obvious things like mount options and filesystem creation it requires bootloader changes, mkinitcpio, dracut, and encryption changes and probably more when you really get into it.

My problem is: Why to make special modifications for GRUB that cannot boot a snapshot if REFInd and `zfsbootmenu` do it or are better? Or user-friendly?

By the way GRUB have also a "recovery" option and the question is: "Can GRUB start a snapshot selecting it with the cursor?"

Besides of the fact that, in contrary to *buntus, arch-based have not multiple kernel to start, but snapshot with zfs.

**Comment by Advantaged** [comment]
> > Besides, some NVMe-manufacturers (Corsair serie [MP510, MP600]) offer the possibility to switch their device to real 4k (4096 bit) PBS/LBS (physical/logical block size) by just one command that could also be implemented in "calamres".
> 
> Not only does this part have essentially nothing to do with ZFS, but this is also not something I would expect an installer to do on a running system. IMHO, it belongs into a kernel CLI option.

The handling of data-carrier is not belong the installer? So is zfs not belong Linux!?

MHO, the installer should be able to detect if data-carrier support native 4k, switch to it automatically (because better and younger than 30 years) and force the formatting applications (gparted & Co.) to format also right, even FAT32 don't support 4k at all XVFAT do it. This should happen already now (or in the past) under GRUB and ext4, but we still carry very old burden.

I know, it's much work, but we should do it now and right, not half-hearted.

Do you know what happen if you install ubuntu on a Samsung NVMe? By the first boot tell the system that the file system is 4k but the kernel registered a 512 bit PBS and will convert the whole FS.
Do you know what happen by installing ubuntu on zfs (even if `zfs-dkms` is installed)? After the first kernel-upgrade GRUB cannot start the machine anymore.

Fact is, I don't want to convince anyone, less to force anything, but just tell about my made experiences.

A user that want install an OS with calamares will neither want to convert any data-carrier PBS/LBS nor make a xvfat-efi partition manually or define another size of swap or move it at end of carrier over the CLI like zfs wish.

Another point is, zfs need an unique identifier for data-carrier (wwn..., meui..., disk-device-by-id) otherwise the fs must be unmounted and remounted for successive data-transfer/OS-installation. These are supported by GRUB, REFIND and ZFSBOOTMENU.

Those all points are to be considered already now, even before starting any zfs implementation.

I'm sure ZFS is the future, now we should make a solid base to reach the necessary acceptance and this, not only under nerds and geeks but for everyone.

**Comment by dalto8** [comment]
> Another example: I read zfs don't "like" Linux-Encryption but LUKS, this should be also considered or investigated as well all other already mentioned points.

ZFS has built-in encryption, there should be little reason to put luks underneath it.

> We should be conscious of the fact that, zfs don't allow dual or multibooting because must be the first partition on the drive.
> If you/we want to remain able to dual multiboot, this must be communicated even/also in calamares. This is the only place where the user really see/read.
zfs doesn't need to be on the first partition of the drive and supports multi-booting just fine.

I am not sure where you are getting your information from but the workstation I am using right now to type this message is a multi-boot zfs machine on a samsung nvme drive with zfs not as the first partition.

**Comment by Advantaged** [comment]
> > Another example: I read zfs don't "like" Linux-Encryption but LUKS, this should be also considered or investigated as well all other already mentioned points.
> 
> ZFS has built-in encryption, there should be little reason to put luks underneath it.
> 
> > We should be conscious of the fact that, zfs don't allow dual or multibooting because must be the first partition on the drive.
> > If you/we want to remain able to dual multiboot, this must be communicated even/also in calamares. This is the only place where the user really see/read.
> > zfs doesn't need to be on the first partition of the drive and supports multi-booting just fine.
> 
> I am not sure where you are getting your information from but the workstation I am using right now to type this message is a multi-boot zfs machine on a samsung nvme drive with zfs not as the first partition.

Than, we don't need anything, everything is perfect, how to, are in the arch-wiki where is written "The installation of ZFS in arch is strait away", we all are dummies and you spend you precious time with us and I ask only one thing: Why?

Or... please post your detailed instruction in order to allow everyone to get ZFS on Linux installed and working, maybe also with booting different snapshots?

Thanks in advance

**Comment by kkofler** [comment]
> Besides of the fact that, in contrary to *buntus, arch-based have not multiple kernel to start, but snapshot with zfs.

This sentence implies that all Arch derivatives use or default to ZFS, which is false. I use one Arch-based distribution, Manjaro ARM on the PinePhone, and it comes on an ext4 `/` partition.

> Do you know what happen if you install ubuntu on a Samsung NVMe? By the first boot tell the system that the file system is 4k but the kernel registered a 512 bit PBS and will convert the whole FS.

That "the kernel registered a 512 bit PBS" part is exactly why the setting belongs into the kernel CLI. Calamares can only work with what the kernel sees.

**Comment by kkofler** [comment]
PS: To elaborate on that last paragraph: The thing is, the kernel will detect the drive early in the bootup and soon after that start mounting partitions as directed by init/systemd. At that point, the block size must be known to the kernel. The most reliable way to make that happen is to pass the block size to the kernel CLI. Of course, the kernel driver (ideally) or at the latest the init system (before mounting the partitions) needs to actually process that CLI parameter, which may or may not be already implemented. But that is the place where it needs to happen, not Calamares.

**Comment by Advantaged** [comment]
> PS: To elaborate on that last paragraph: The thing is, the kernel will detect the drive early in the bootup and soon after that start mounting partitions as directed by init/systemd. At that point, the block size must be known to the kernel. The most reliable way to make that happen is to pass the block size to the kernel CLI. Of course, the kernel driver (ideally) or at the latest the init system (before mounting the partitions) needs to actually process that CLI parameter, which may or may not be already implemented. But that is the place where it needs to happen, not Calamares.

### NVMe
1. `nvme-cli` must be installed by default on all live systems like others data-carrier-modification application too, or:
` sudo pacman -S nvme-cli`
2. Detect if nvme is present:
`nvme --list`
3. Detect if nvme support 4k:
```
nvme id-ns /dev/nvme0n1 -H | grep LBA
```
> output:

```
LBA Format  0 : Metadata Size: 0   bytes - Data Size: 512 bytes - Relative Performance: 0x2 Good (in use)
LBA Format  1 : Metadata Size: 0   bytes - Data Size: 4096 bytes - Relative Performance: 0x1 Better
```
4. format it to "LBA Format 1" with:
```
nvme format /dev/nvme0n1 -l 1
```
> output:
```
Success formatting namespace:1
````
5. recheck
```
nvme id-ns /dev/nvme0n1 -H | grep LBA
```
> output:

```
LBA Format  0 : Metadata Size: 0   bytes - Data Size: 512 bytes - Relative Performance: 0x2 Good 
LBA Format  1 : Metadata Size: 0   bytes - Data Size: 4096 bytes - Relative Performance: 0x1 Better (in use)
```
Note: Data-carrier not supporting "namespace"-formatting has only one output (LBA format 0). Samsung DC (data-carrier)  have PBS = 512, LBS = 512e, those DC working better in Linux after `secure-erase` indeed PBS and LBS are both set to 512 bit.

### Message by detecting DC is convertible if a reboot is necessary
The drive you choose is convertible to 4096 bit instead of (old) 512 bit. This is called 4k-sector-size.

In order the conversion is recognized by kernel, a reboot is necessary.

Convert now: [O] YES convert [O] NOT convert and proceed

After conversion (if any) and check are done offer to reboot.

## 4k Format (commands)
```
mkfs.vfat -F32 -s 2 -S 4096 -v /dev/nvme0n1p1

mkfs.ext4 -F -b 4096 -F /dev/nvme0n1p2

mkswap /dev/nvme0n1p3
```
Note: I don't know how to make swap with 4096 bit. EFI partition type remain `ef00` formatting is `vfat`. You can set the variable e.g.:  `DISK=/dev/disk/by-id/wwn-0x500277a4100c4e21` or "serial number" or "UUID" by selecting DC.
Due most changeable UUID is recommended to use "serial-number" or "disk/by-id" also cause of uncertain UUID befor partitioning and formatting.
Example partitioning (using "disk-variable"):
```
sgdisk -n 1:1M:+1G -t 1:ef00 -c 1:EFI-0001 $DISK

sgdisk -n 2:0:+428G -t 2:8300 -c 2:arch-0001 $DISK

sgdisk -n 3:0:+36G -t 3:8300 -c 3:swap-0001 $DISK
```
Example formatting (using "disk-variable"):

```
mkfs.vfat -F32 -s 2 -S 4096 -v $Disk-part1

mkfs.ext4 -F -b 4096 -F &DISK-part2

mkswap $DISK-part3
```
The graphical formatting application are also vintage-style indeed the `GiB` selection is not available, only `MiB` is there, this is uncomfortable.
By using zfs-raid and assure any "spare-disk" any manufacturer or model to fit always and let some reserve at end (even for SATA-SSD and HDD), must be a solution to reduce the maximal capacity of DC. The actual graphical partition application is not "state of the art", True-NAS make it right, Linux don't.

I don't know why swap don't take anymore last partition on DC as before SSD become popular, anyway, for HDD and ZFS is essential to reoccupy latest partition.

If we have to reboot after PBS/LBS-conversion we have to, but we cannot still remain on 30-40 years old block-size, even all manufacturer specify the DC-speed at 4k even if deliver the DC in 512 or 512e. All manufacturer delivering non 4k-option or 4k-default want to support, at any cost, OS not supporting or understanding 4k.

**Comment by dalto8** [comment]
I am deeply confused as to what optimizing the sector size of the physical disk has to do with adding zfs support.  If there is a desire for Calamares to try to optimize the hardware, shouldn't that be a separate issue?

**Comment by Advantaged** [comment]
> > please post your detailed instruction in order to allow everyone to get ZFS on Linux installed and working
> 
> The steps to get zfs working on Linux varies per distribution. Even if I was motivated/inclined to provide such steps, I don't think an issue request for Calamares would be an appropriate place to create such a tutorial.


you can post me in VK.com or to gmail, my first name is tony."what you see here = adv..." at....
My preferred distro is Artix Plasma OpenRC or Majaro plasma (for children notebooks).
If you post here, you have a github-account, like me, have you there something? or can you contact me on my github?

> > Than, we don't need anything, everything is perfect, how to, are in the arch-wiki where is written "The installation of ZFS in arch is strait away", we all are dummies and you spend you precious time with us and I ask only one thing: Why?
> 
> I have no idea what this about or what you are referring to here. I simply pointed out that some of the information you stated is not accurate which seems important if the end result of this issue is to have zfs support in calamares.


I explain only the experience I have done, I dismiss *ubuntus and Samsung, hence I cannot provide smoking guns, but if you can.. do it please but not here as PM, OK?
> > This sentence implies that all Arch derivatives use or default to ZFS, which is false.
> 
> I think the point he was trying to make is that Arch-based distros don't keep prior kernels around like many distros do so if you want to rollback to a prior kernel a snapshot is a way to do that. That being said, I don't think that implementing snapshot booting needs support in Calamares. It seems like it should be handled by the distro.

Keeping older kernels around is true by debian and derivates like *buntus, devuan, mint, linux ultimate, etc. and that I mean too.


ZFS should also be handled by distro by adding it in the kernel, but they don't too.
Arch and manjaro kernels are newer than zfs-packages and own ISO-creating-applications, hence, both don't work at all.
Honestly speaking!? they don't want to support neither own iso-creation apps nor zfs.
Even refind, artix-linux has a special one with artix-logo, is not included or choosable by the installation with calamres, but artix use older kernels than arch and manjaro, hence should be "easier" to get calamares for artix and vorall with openrc (see possible special handling under systemD).
Arto-linux have own written calamares with many options not existing in other distros at all.
If the best method to start zfs is refind + zfsbootmenu and calamres don't implement it or wait for the distro-s... good night.

Of course we have to talk with distros developers and/or maintainers because... you can develop the best calamares of all times, but if the maintainer or developer drop it and use the older one or disable the new functions... this is equal to a hole in the water.

**Comment by kkofler** [comment]
So, I have 2 points to make:
1. Please file separate issue reports for separate actual issues. This report mixes at least 2 unrelated issues: support for the ZFS file system and support for native 4k blocks. They need to be filed separately. Catch-all issues will just eventually be closed because the developers lose track of what is actually requested in them.
2. Calamares is primarily a toolkit for distribution developers allowing them to offer an installer for their distribution. The features Calamares supports are primarily driven by distributions' needs. It is not the job of Calamares to work around limitations in the distribution or to offer features that do not belong in the installer.

**Comment by Advantaged** [comment]
> I am deeply confused as to what optimizing the sector size of the physical disk has to do with adding zfs support. If there is a desire for Calamares to try to optimize the hardware, shouldn't that be a separate issue?

I cannot answer anymore to your suspicions because out of topic. Fact is, if I buy newest hardware and want to use most modern features, I have to partition and format manually. This isn't fold of calamares developers.

Sorry kkofler, you are right, this can confuse the developer, I'm sorry.

I thought calamres is the installer and just missing the options for zfs and all other things like 4k block size, partitioning, formatting, etc..

If I remember well, even in other distros, the scripts are done by distros, the calamares offer to show the selectable options in the gui through the  tools acting as interpreter between scripts and guy on front of display and maybe more.

I know now approximately what's the situation, nevertheless the topics I addressed have to be discussed or considered even if in another space.

Basically and finally no ZFS support/implementation until the distro developer put it in the ISO with the right kernel-version or modify the kernel same implementing ZFS in kernel-package/drivers and do it by each kernel-version either to check if kernel-version is compatible with zfs-versions or implementing the drivers in the kernel.

Sorry again if I cause some confusion here.

**Comment by kmath313** [comment]
I have almost no understanding of how Calamares does it's business, but I'm happy to start learning. I have used ZFS on root for both Arch (Arcolinux to be exact) and Debian, in the case of the Arch build I used the ZFS module specifically tied to the kernel version but for my Debian install I am using ZFS DKMS. 

Whilst I know there potential issues with DKMS it does de-couple the ZFS module version from the kernel version, so that could be what any live ISO uses to conduct the install. But Calamares will be reliant on the distro packager adding the ArchZFS repo to pacman.conf to do any of this.

The OpenZFS page has a very detailed set of instructions for installing Arch on ZFS, I am attempting to put together a script that will run the required commands to get a basic TTY-only install of Arch on my specific hardware. If I get it working smoothly I would be happy to share it here and have the Calamares guys pull the bits needed to make it work.

For the 4k vs 512 sector size Calamares could run the nvme commands listed above to determine the actual sector size and then pass ZFS either an ashift of 9 (512b sectors) or 12 (4k), but I would recommend just passing ashift=12 as there is no penalty associated with using that setting with a 512b sector drive.

The OpenZFS guide has individual datasets set up for each individual user of the system - I would suggest that we just set a single dataset for /home (like is currently done with a BTRFS install) otherwise there is a bunch of extra commands to run just to add a new user. 

Using GRUB to boot is fine - the OpenZFS tutorial shows the specific zpool create command to hold /boot.

I hope some of this helps the progression towards implementing ZFS on Root in Calamares.

**Comment by dalto8** [comment]
I spent some time on this today.  One of the challenges is that the support in kpmcore is quite limited.  That means there are going to be tons of `if ( fs->type() == FileSystem::Type::Zfs )` or similar checks all over the place to make this work.  There will be many places where kpmcore will need to bypassed.

**Comment by Advantaged** [comment]
I found the latest (march) [instructions](https://github.com/openzfs/openzfs-docs/blob/91d28894a74a19f1a528227925ceb5ff45dc236a/docs/Getting%20Started/Arch%20Linux/Artix%20Linux%20Root%20on%20ZFS.rst) for installing 'Artix OpenRC' on ZFS.

There are many differences to 'Arch-Linux-Wiki' but even much more details e.g.: the structure of data-carrier, how to choose the right kernel for the right ZFS-versions, all the problems with GRUB and alternatives, etc., etc.. 

[Erik Dubois](https://www.youtube.com/c/ErikDubois/about) have a lot of experience,

**Comment by kmath313** [comment]
@Advantaged yup! Those are the instructions I have been using - well I pulled from the website (https://openzfs.github.io/openzfs-docs/Getting%20Started/Arch%20Linux/Arch%20Linux%20Root%20on%20ZFS.html). 

Is there a way for kpmcore to be bypassed if the user selects ZFS? That way we could run the sgdisk commands to clear, the partition and then run the zpool/zfs commands to create the pool and datasets.

**Comment by dalto8** [comment]
Just my opinion, but I don't think the challenges with adding this feature have anything to do with needing information or tutorials.  If the person(s) implementing this functionality need to read a tutorial on how to setup zfs it is going to be a tough implementation because there are a lot of little details that an installer has to handle that are not an issue in a single install to a specific distro.

It seems like there are two things that are needed to get this implemented in Calamares.
1. The Calamares team has to be open to the complexity that it would add to the code base.  It will require changes to quite a few modules and a bunch of specific conditional logic in many functions.
2. Someone has to be be willing to invest the time and do the work.  If you have a decent working knowledge of zfs, it isn't conceptually difficult but it is still a pretty big investment in time to get all the code written and tested.

I implemented the zfs support in Manjaro-Architect installer(now defunct) so I have a pretty good idea of what it would take to implement here.  However, before I start investing a ton of time on it I would want to know that the dev team would interested in merging it because it will require changes all over the code base.

**Comment by kmath313** [comment]
@Advantaged - I haven't asked stikonas but I did do some digging into kpmcore and it uses sfdisk. It looks like we could set up the partition structure as per the OpenZFS instructions using sfdisk. So that is one (very minor) issue "solved"....

You are correct the sfdisk/sgdisk commands simply set the partition boundaries and it only gets formatted when the zpool command is issued. 

GRUB does work, but you do need to set it up to do so. So it definitely would be a case of if(fs=zfs).... so that additional text could be added to the grub.conf file prior to building the grub.cfg. The other options are reFind - but you would need to load additional drivers to deal with ZFS or the last option is systemd-boot. That said once the workarounds for GRUB have been added to the grub.conf I never had an issue with GRUB not updating correctly when I updated my kernel or updated the GRUB package.

I agree - GRUB should be able to read from /dev/disk/by-id or /dev/disk/by-UUID. 

A question for the other Calameres developers or GRUB experts - would there be any issues for non-ZFS users if Calameres set 'export ZPOOL_VDEV_NAME_PATH=YES' by default during the GRUB install process?

Once the canonical path problem is solved then there will be a if (fs=zfs).. statement to replace the standard grub-probe command with the zdb one to ensure that the root pool is specified correctly in the grub kernel commandline.

**Comment by Tony-Goat** [comment]
> Comment created by sb56637:
> 
> The automated modes should be good enough. I wonder if code could be used from the Antergos Cnchi installer, which does support ZFS root installations: https://github.com/Antergos/Cnchi

I was actually curious in Calamares and ZFS and contributing something to both the conversation and the project.

I broke down Cnchi and took notes on how it works. This may have changed because the version specifically referenced in Cnchi's source code is 0.6.5.4, which is several versions old. [I put my notes in a gist here.](https://gist.github.com/Tony-Goat/5b792063ebce8a4f2c68184bbad44b44) I don't know if this process is anywhere near the same as Calamares.

The meat of the install process is in src/pages/zfs_manager.py:setup(), but most of the stuff that actually makes it work to boot is littered through the files in the src/installation folder. This setup differs from Ubuntu in that the boot partition is just an ext4 partition with an initrd in it, which sets up the main ZFS root, instead of having a bpool and an rpool. This might be preferable to some users.

**Comment by kmath313** [comment]
@Tony-Goat - It looks like Cnchi was setting up swap on a ZFS zvol - the last I saw that was still not a recommended set up.. What I didn't see in your notes is what datasets are created during the install, we don't have to create datasets but I think it would be beneficial as when the user (or us) set up snapshots they don't snapshot folders that don't to be snapshotted eg /var/spool.

I did like that it gave the option for /home to be on its own pool but did Cnchi allow for that pool to be created on a different disk in the system? Or was it just creating a separate zpool partition on the single drive?

**Comment by kmath313** [comment]
I think the main reason for the separate bpool is simply because GRUB2 can't handle all the features of ZFS. Whilst we can solve this by using a different bootloader I think this is adding more complications that we don't need to worry about.... So I guess we should put it to all in this thread that our partition structure would look something like what you proposed.

**Comment by Advantaged** [comment]
@dalto8 
This time I agree 100% with you.
The possibility to choose is the key of success even if we have to touch topics like bootloader, partitioning, 4k-conversion, etc.

It's time to stop "running" behind other OS-, Hardware- and Software-Manufacturer, try to unify human-resources and make the best choice for the future.

What i said, OpenZFS is already 10 years old, no Linux-OS implemented it correctly or at all. Can you imagine what, Linux-Users want ZFS, will do? I'm also thinking of switching to FreeBSD. Inside FreeBSD, than, I can make jails or VMs with each OS I want.

We (in the Linux-World) are always late, this time very late, we should take every possible help to succeed.

**Comment by kmath313** [comment]
@Advantaged from my quick reading (and somewhat limited knowledge) it looks like kpmcore would be able to issue commands to establish a partition structure that would support ZFS on root. So I might be wrong (and probably are....).

**Comment by dalto8** [comment]
>  it looks like kpmcore would be able to issue commands to establish a partition structure that would support ZFS on root.

It will not in it's current form but that is not of particular concern in this case.

**Comment by Advantaged** [comment]
@Tony-Goat 
Calamares doesn't do this, doesn't do that, who is Calamares? Are you Calamares perhaps?
I know only, Arco-Linux toke Arch-Linux, put and enhanced Calamares on it and the results are very good.
The only disadvantage, I see in Arco, is "SystemD".
The only possible alternative to this fact is, I'm talking with the wrong people. For sure, I can't talk with Calamares but with humans.
If you mean calamares should be an universal tool, then we need a kind of mini-system with all necessary tools and additional partition to store ISO's, e.g.:
> Part 1: EFI with refind, all drivers including iso, zfs, vfat, gpu, etc.
> Part 2: Calamares with all necessary tools that by each start scan part 3 for ISO's
> Part 3: here store one or more ISO's Calamares will kick off

Anyway we have to github a rEFIInd with zfs-drivers (or all drivers) e.g. refind-zfs (or refind-ad) and by the way fix some small inconveniences like right phat to icons, add missing icons, replace default UUID with `/dev/disk/by-id/` ('wwn' and/or 'nvme-eui') and maybe also offer different backgrounds. I sent e-mail to refind-coder and drivers-coder, we will see. 

@kmath313 
II don't know `kpmcore`, the SE (search engine) tell me that's a Calamares dependencies, has to do with Kpart and have some [little problems](https://github.com/calamares/calamares/issues/1724) by setting 'flags' properly.

@dalto8 
exactly, kpmcore (KDE partition manager) don't contains 'vfat' out of the box and cannot formate EFI-partition or any other partition with 4k. Due the fact that ZFS requires, as kmath313 already wrote, "partitions boundaries" first, than build zpool and only after them (maybe) format the (4) "partitions"... kpmcore is not a relevant issue anymore. I know, this is a pain but, again, we cannot hope/wait someone else build/[fix](https://github.com/calamares/calamares/issues/1724) if and when we want. 

But, before we start to recycle again the topics (kernel, licenses, partitioning, etc.) let give a look at the facts and correlated questions:
1. F: Each distribution use a modified kernel (e.g. Arch, Artix, Manjaro, Debian, *ubuntu), so they can add the 'zfs-module' but don't want to.
2. F: The worst on installations-tools is the very restrictive handling of data carriers, this is valid for Ubiquity as well for "our" Calamares. At least, even Ubiquity use for ZFS-installation a Python-Script and not 'kpmcore'. 
3. F: None of Graphical-Partitioning-Tools make a difference between partitioning and formatting (crucial not only for ZFS), no graphical-partitioning-tool or even OS can make (out of the box) `vfat` without manual install of additional packages and/or module. None of them let partitioning with 4k blocks, even the expert user don't know which kernel-module are loaded and which not. None of them show the capacity
4. Q: Do we have a works-plane, or what's your plane, can you tell us? Maybe we can overtake some parts of them!
5. Q: Have we to care/choose first Data-Carrier-Handling with reliable tools before we proceed?
6. Q: Is formatting with `vfat`, `ext4` and `swap` necessary or just make "partitions-boundaries"?
7. Q: Should ZFS spans over all partitions or only over some of them?
8. Q: Is possible to insert ZFS-password in the script (like by Ubiquity) with the result only one time password by booting final system or will be automatically uses the password of desktop-manager?
9. Q: Will be there a choice to select to encrypt the System, and which encryption type?
10.  Q: Will be there the choice to select to make snapshots?
11. Q: In case snapshot can/will be enabled, will be a tool there a tool (e.g. `zbm` (ZFS-Boot-Menu)) to select any snapshot and start it?
12. Q: For which distributions do you plane to use "Calamares-ZFS", arch-derivates or more?

**Comment by adriaandegroot** [comment]
- rEFInd support (thanks to demmm) is available through the *refind* module in calamares-extensions (this can live independently of what the rest of Calamares does)
- ZFS support (thanks to dalto8) is available through the *zfs* module in calamares (this needs pretty tight integration with *partition* and *mount* and other modules in core, anyway)
